### PR TITLE
[Platform API] Add test cases for new API methods

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
### Description of PR

Summary:
During SONiC physical entity MIB feature development, a few new platform APIs were added and they need regression test cases to verify them.

Depends on https://github.com/Azure/sonic-platform-common/pull/134

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### What is the motivation for this PR?

Add test cases and test support for all newly added platform API introduced by SONiC physical entity MIB feature

#### How did you do it?

Add new API test support for chassis, fan, fan drawer, psu, sfp and add new test cases for them.

#### How did you verify/test it?

Manually run the regression

#### Any platform specific information?

N/A

#### Supported testbed topology if it's a new test case?

N/A

### CODE REVIEW

File: tests/platform_tests/api/test_psu.py [hunk 1]
```diff
@@ -200,3 +200,33 @@ def test_led(self, duthost, localhost, platform_api_conn):
...
+            replaceable = psu.is_replaceable(platform_api_conn, psu_id)
+            self.expect(replaceable is not None, "Failed to perform is_replaceable for psu id {}".format(psu_id))
+            self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for psu id [}".format(psu_id))
+        self.assert_expectations()
+
+    def test_thermals(self, platform_api_conn):
+        for psu_id in range(self.num_psus):
+            try:
+                num_thermals = int(psu.get_num_thermals(platform_api_conn, psu_id))
+            except:
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:221]: To catch all exceptions, please use `except Exception:`

File: tests/platform_tests/api/test_sfp.py [hunk 2]
```diff
@@ -423,3 +423,33 @@ def test_power_override(self, duthost, localhost, platform_api_conn):
...
+            replaceable = sfp.is_replaceable(platform_api_conn, i)
+            self.expect(replaceable is not None, "Failed to perform is_replaceable for sfp {}".format(i))
+            self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for sfp [}".format(i))
+        self.assert_expectations()
+
+    def test_thermals(self, platform_api_conn):
+        for i in range(self.num_sfps):
+            try:
+                num_thermals = int(sfp.get_num_thermals(platform_api_conn, i))
+            except:
```
Comment (jleveque) [inline tests/platform_tests/api/test_sfp.py:445]: To catch all exceptions, please use `except Exception:`

File: tests/platform_tests/api/test_psu.py [hunk 3]
```diff
@@ -200,3 +200,33 @@ def test_led(self, duthost, localhost, platform_api_conn):
                         self.expect(color == color_actual, "Status LED color incorrect (expected: {}, actual: {}) from PSU {}".format(color, color_actual, psu_id))
         self.assert_expectations()
 
+    def test_get_position_in_parent(self, platform_api_conn):
+        for psu_id in range(self.num_psus):
+            position = psu.get_position_in_parent(platform_api_conn, psu_id)
+            self.expect(position is not None, "Failed to perform get_position_in_parent for psu id {}".format(psu_id))
+            self.expect(isinstance(position, int), "Position value must be an integer value for psu id {}".format(psu_id))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:207]: The second `self.expect()` call is irrelevant if the first fails, so it should be nested using `if self.expect()` on the first. See lines 194-195 as an example.

File: tests/platform_tests/api/test_psu.py [hunk 4]
```diff
@@ -200,3 +200,33 @@ def test_led(self, duthost, localhost, platform_api_conn):
...
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for psu_id in range(self.num_psus):
+            replaceable = psu.is_replaceable(platform_api_conn, psu_id)
+            self.expect(replaceable is not None, "Failed to perform is_replaceable for psu id {}".format(psu_id))
+            self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for psu id [}".format(psu_id))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:214]: The second `self.expect()` call is irrelevant if the first fails, so it should be nested using `if self.expect()` on the first. See lines 194-195 as an example.

File: tests/platform_tests/api/test_chassis.py [hunk 5]
```diff
@@ -381,3 +381,15 @@ def test_get_watchdog(self, duthost, localhost, platform_api_conn):
     def test_get_eeprom(self, duthost, localhost, platform_api_conn):
         eeprom = chassis.get_eeprom(platform_api_conn)
         pytest_assert(eeprom is not None, "Failed to retrieve system EEPROM")
+
+    def test_get_position_in_parent(self, platform_api_conn):
+        position = chassis.get_position_in_parent(platform_api_conn)
+        self.expect(position is not None, "Failed to perform get_position_in_parent")
+        self.expect(isinstance(position, int), "Position value must be an integer value")
```
Comment (jleveque) [inline tests/platform_tests/api/test_chassis.py:388]: The second `self.expect()` call is unnecessary if the first fails, so it should be nested using `if self.expect()` on the first. See lines 364-365 as an example.

File: tests/platform_tests/api/test_chassis.py [hunk 6]
```diff
@@ -381,3 +381,15 @@ def test_get_watchdog(self, duthost, localhost, platform_api_conn):
...
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        replaceable = chassis.is_replaceable(platform_api_conn)
+        self.expect(replaceable is not None, "Failed to perform is_replaceable")
+        self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value")
```
Comment (jleveque) [inline tests/platform_tests/api/test_chassis.py:394]: The second `self.expect()` call is unnecessary if the first fails, so it should be nested using `if self.expect()` on the first. See lines 364-365 as an example.

File: tests/platform_tests/api/test_chassis_fans.py [hunk 7]
```diff
@@ -221,3 +221,17 @@ def test_set_fans_led(self, duthost, localhost, platform_api_conn):
                             color, color_actual, i))
 
         self.assert_expectations()
+
+    def test_get_position_in_parent(self, platform_api_conn):
+        for i in range(self.num_fans):
+            position = fan.get_position_in_parent(platform_api_conn, i)
+            self.expect(position is not None, "Failed to perform get_position_in_parent for fan {}".format(i))
+            self.expect(isinstance(position, int), "Position value must be an integer value for fan {}".format(i))
```
Comment (jleveque) [inline tests/platform_tests/api/test_chassis_fans.py:229]: The second `self.expect()` call is unnecessary if the first fails, so it should be nested using `if self.expect()` on the first. See lines 213-214 as an example.

File: tests/platform_tests/api/test_chassis_fans.py [hunk 8]
```diff
@@ -221,3 +221,17 @@ def test_set_fans_led(self, duthost, localhost, platform_api_conn):
...
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_fans):
+            replaceable = fan.is_replaceable(platform_api_conn, i)
+            self.expect(replaceable is not None, "Failed to perform is_replaceable for fan {}".format(i))
+            self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for fan [}".format(i))
```
Comment (jleveque) [inline tests/platform_tests/api/test_chassis_fans.py:236]: The second `self.expect()` call is irrelevant if the first fails, so it should be nested using `if self.expect()` on the first. See lines 213-214 as an example.

File: tests/platform_tests/api/test_sfp.py [hunk 9]
```diff
@@ -423,3 +423,33 @@ def test_power_override(self, duthost, localhost, platform_api_conn):
             if self.expect(power_override is not None, "Unable to retrieve transceiver {} power override data".format(i)):
                 self.expect(power_override is False, "Transceiver {} power override data is incorrect".format(i))
         self.assert_expectations()
+
+    def test_get_position_in_parent(self, platform_api_conn):
+        for i in range(self.num_sfps):
+            position = sfp.get_position_in_parent(platform_api_conn, i)
+            self.expect(position is not None, "Failed to perform get_position_in_parent for sfp {}".format(i))
+            self.expect(isinstance(position, int), "Position value must be an integer value for sfp {}".format(i))
```
Comment (jleveque) [inline tests/platform_tests/api/test_sfp.py:431]: The second `self.expect()` call is irrelevant if the first fails, so it should be nested using `if self.expect()` on the first. See lines 416-417 as an example.

File: tests/platform_tests/api/test_sfp.py [hunk 10]
```diff
@@ -423,3 +423,33 @@ def test_power_override(self, duthost, localhost, platform_api_conn):
...
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_sfps):
+            replaceable = sfp.is_replaceable(platform_api_conn, i)
+            self.expect(replaceable is not None, "Failed to perform is_replaceable for sfp {}".format(i))
+            self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for sfp [}".format(i))
```
Comment (jleveque) [inline tests/platform_tests/api/test_sfp.py:438]: The second `self.expect()` call is irrelevant if the first fails, so it should be nested using `if self.expect()` on the first. See lines 416-417 as an example.

File: tests/platform_tests/api/test_thermal.py [hunk 11]
```diff
@@ -206,3 +206,17 @@ def test_set_high_threshold(self, duthost, localhost, platform_api_conn):
                     self.expect(temperature == 80,
                                 "Thermal {} high threshold {} is not matching the set value {}".format(i, temperature, high_temperature))
         self.assert_expectations()
+
+    def test_get_position_in_parent(self, platform_api_conn):
+        for i in range(self.num_thermals):
+            position = thermal.get_position_in_parent(platform_api_conn, i)
+            self.expect(position is not None, "Failed to perform get_position_in_parent for thermal {}".format(i))
+            self.expect(isinstance(position, int), "Position value must be an integer value for thermal {}".format(i))
```
Comment (jleveque) [inline tests/platform_tests/api/test_thermal.py:214]: The second `self.expect()` call is irrelevant if the first fails, so it should be nested using `if self.expect()` on the first. See lines 200-201 as an example.

File: tests/platform_tests/api/test_thermal.py [hunk 12]
```diff
@@ -206,3 +206,17 @@ def test_set_high_threshold(self, duthost, localhost, platform_api_conn):
...
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_thermals):
+            replaceable = thermal.is_replaceable(platform_api_conn, i)
+            self.expect(replaceable is not None, "Failed to perform is_replaceable for thermal {}".format(i))
+            self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for thermal [}".format(i))
```
Comment (jleveque) [inline tests/platform_tests/api/test_thermal.py:221]: The second `self.expect()` call is irrelevant if the first fails, so it should be nested using `if self.expect()` on the first. See lines 200-201 as an example.

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — APPROVED** [review]

File: tests/platform_tests/api/test_fan_drawer.py [hunk 13]
```diff
@@ -154,3 +154,17 @@ def test_set_fan_drawers_led(self, duthost, localhost, platform_api_conn):
                             color, color_actual, i))
 
         self.assert_expectations()
+
+    def test_get_position_in_parent(self, platform_api_conn):
+        for i in range(self.num_fan_drawers):
+            position = fan_drawer.get_position_in_parent(platform_api_conn, i)
+            self.expect(position is not None, "Failed to perform get_position_in_parent for fan drawer {}".format(i))
+            self.expect(isinstance(position, int), "Position value must be an integer value for fan drawer {}".format(i))
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_fan_drawer.py:162]: The second self.expect() call is irrelevant if the first fails, so it should be nested using if self.expect() on the first.

File: tests/platform_tests/api/test_fan_drawer.py [hunk 14]
```diff
@@ -154,3 +154,17 @@ def test_set_fan_drawers_led(self, duthost, localhost, platform_api_conn):
...
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_fan_drawers):
+            replaceable = fan_drawer.is_replaceable(platform_api_conn, i)
+            self.expect(replaceable is not None, "Failed to perform is_replaceable for fan drawer {}".format(i))
+            self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for fan drawer [}".format(i))
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_fan_drawer.py:169]: The second self.expect() call is irrelevant if the first fails, so it should be nested using if self.expect() on the first.

File: tests/platform_tests/api/test_psu.py [hunk 15]
```diff
@@ -200,3 +200,33 @@ def test_led(self, duthost, localhost, platform_api_conn):
...
+
+    def test_thermals(self, platform_api_conn):
+        for psu_id in range(self.num_psus):
+            try:
+                num_thermals = int(psu.get_num_thermals(platform_api_conn, psu_id))
+            except Exception:
+                pytest.fail("num_thermals is not an integer")
+
+            thermal_list = psu.get_all_thermals(platform_api_conn, psu_id)
+            pytest_assert(thermal_list is not None, "Failed to retrieve thermals")
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_psu.py:225]: Please add the psu_id to the pytest_assert.

```suggestion
            pytest_assert(thermal_list is not None, "Failed to retrieve thermals for psu {}".format(psu_id))
```

File: tests/platform_tests/api/test_psu.py [hunk 16]
```diff
@@ -200,3 +200,33 @@ def test_led(self, duthost, localhost, platform_api_conn):
...
+            pytest_assert(isinstance(thermal_list, list) and len(thermal_list) == num_thermals, "Thermals appear to be incorrect")
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_psu.py:226]: Please add the psu_id to the pytest_assert.

File: tests/platform_tests/api/test_sfp.py [hunk 17]
```diff
@@ -423,3 +423,33 @@ def test_power_override(self, duthost, localhost, platform_api_conn):
...
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_sfps):
+            replaceable = sfp.is_replaceable(platform_api_conn, i)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for sfp {}".format(i)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for sfp [}".format(i))
+        self.assert_expectations()
+
+    def test_thermals(self, platform_api_conn):
+        for i in range(self.num_sfps):
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_sfp.py:442]: suggest changing this i to sfp_id for unifomity

File: tests/platform_tests/api/test_sfp.py [hunk 18]
```diff
@@ -423,3 +423,33 @@ def test_power_override(self, duthost, localhost, platform_api_conn):
...
+            try:
+                num_thermals = int(sfp.get_num_thermals(platform_api_conn, i))
+            except Exception:
+                pytest.fail("num_thermals is not an integer")
+
+            thermal_list = sfp.get_all_thermals(platform_api_conn, i)
+            pytest_assert(thermal_list is not None, "Failed to retrieve thermals")
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_sfp.py:449]: Please add the sfp_id to the pytest_assert.

File: tests/platform_tests/api/test_sfp.py [hunk 19]
```diff
@@ -423,3 +423,33 @@ def test_power_override(self, duthost, localhost, platform_api_conn):
...
+            pytest_assert(isinstance(thermal_list, list) and len(thermal_list) == num_thermals, "Thermals appear to be incorrect")
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_sfp.py:450]: Please add the sfp_id to the pytest_assert.

**Review by vdahiya12 — CHANGES_REQUESTED** [review]

File: tests/platform_tests/api/test_chassis_fans.py [hunk 20]
```diff
@@ -221,3 +221,17 @@ def test_set_fans_led(self, duthost, localhost, platform_api_conn):
...
+            position = fan.get_position_in_parent(platform_api_conn, i)
+            if self.expect(position is not None, "Failed to perform get_position_in_parent for fan {}".format(i)):
+                self.expect(isinstance(position, int), "Position value must be an integer value for fan {}".format(i))
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_fans):
+            replaceable = fan.is_replaceable(platform_api_conn, i)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for fan {}".format(i)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for fan [}".format(i))
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_chassis_fans.py:236]: type 
```suggestion
                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for fan {}".format(i))
```

File: tests/platform_tests/api/test_fan_drawer.py (hunk 14 above)
Comment (vdahiya12) [inline tests/platform_tests/api/test_fan_drawer.py:169]: please fix the typo
```suggestion
            self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for fan drawer {}".format(i))
```

File: tests/platform_tests/api/test_psu.py [hunk 21]
```diff
@@ -200,3 +200,33 @@ def test_led(self, duthost, localhost, platform_api_conn):
...
+            position = psu.get_position_in_parent(platform_api_conn, psu_id)
+            if self.expect(position is not None, "Failed to perform get_position_in_parent for psu id {}".format(psu_id)):
+                self.expect(isinstance(position, int), "Position value must be an integer value for psu id {}".format(psu_id))
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for psu_id in range(self.num_psus):
+            replaceable = psu.is_replaceable(platform_api_conn, psu_id)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for psu id {}".format(psu_id)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for psu id [}".format(psu_id))
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_psu.py:214]: please fix the typo
```suggestion
                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for psu id {}".format(psu_id))
```

File: tests/platform_tests/api/test_sfp.py [hunk 22]
```diff
@@ -423,3 +423,33 @@ def test_power_override(self, duthost, localhost, platform_api_conn):
...
+            position = sfp.get_position_in_parent(platform_api_conn, i)
+            if self.expect(position is not None, "Failed to perform get_position_in_parent for sfp {}".format(i)):
+                self.expect(isinstance(position, int), "Position value must be an integer value for sfp {}".format(i))
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_sfps):
+            replaceable = sfp.is_replaceable(platform_api_conn, i)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for sfp {}".format(i)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for sfp [}".format(i))
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_sfp.py:438]: please fix this typo`
``suggestion
                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for sfp {}".format(i))
```

File: tests/platform_tests/api/test_thermal.py [hunk 23]
```diff
@@ -206,3 +206,17 @@ def test_set_high_threshold(self, duthost, localhost, platform_api_conn):
...
+            position = thermal.get_position_in_parent(platform_api_conn, i)
+            if self.expect(position is not None, "Failed to perform get_position_in_parent for thermal {}".format(i)):
+                self.expect(isinstance(position, int), "Position value must be an integer value for thermal {}".format(i))
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_thermals):
+            replaceable = thermal.is_replaceable(platform_api_conn, i)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for thermal {}".format(i)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for thermal [}".format(i))
```
Comment (vdahiya12) [inline tests/platform_tests/api/test_thermal.py:221]: please fix this typo
```suggestion
                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for thermal {}".format(i))
```

**Review by vdahiya12 — CHANGES_REQUESTED** [review]

**Review by vdahiya12 — APPROVED** [review]

File: tests/platform_tests/api/test_chassis_fans.py [hunk 24]
```diff
@@ -221,3 +221,17 @@ def test_set_fans_led(self, duthost, localhost, platform_api_conn):
...
+            if self.expect(position is not None, "Failed to perform get_position_in_parent for fan {}".format(i)):
+                self.expect(isinstance(position, int), "Position value must be an integer value for fan {}".format(i))
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_fans):
+            replaceable = fan.is_replaceable(platform_api_conn, i)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for fan {}".format(i)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for fan {}".format(i))
+        self.assert_expectations()
```
Comment (sujinmkang) [inline tests/platform_tests/api/test_chassis_fans.py:237]: same comments as test_chassis.py.

File: tests/platform_tests/api/test_psu.py [hunk 25]
```diff
@@ -200,3 +200,33 @@ def test_led(self, duthost, localhost, platform_api_conn):
...
+                self.expect(isinstance(position, int), "Position value must be an integer value for psu id {}".format(psu_id))
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for psu_id in range(self.num_psus):
+            replaceable = psu.is_replaceable(platform_api_conn, psu_id)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for psu id {}".format(psu_id)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for psu id {}".format(psu_id))
+        self.assert_expectations()
+
```
Comment (sujinmkang) [inline tests/platform_tests/api/test_psu.py:216]: same comments as test_chassis.py.

File: tests/platform_tests/api/test_sfp.py [hunk 26]
```diff
@@ -423,3 +423,33 @@ def test_power_override(self, duthost, localhost, platform_api_conn):
...
+                self.expect(isinstance(position, int), "Position value must be an integer value for sfp {}".format(i))
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for sfp_id in range(self.num_sfps):
+            replaceable = sfp.is_replaceable(platform_api_conn, sfp_id)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for sfp {}".format(sfp_id)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for sfp {}".format(sfp_id))
+        self.assert_expectations()
+
```
Comment (sujinmkang) [inline tests/platform_tests/api/test_sfp.py:440]: same comments as test_chassis.py.

File: tests/platform_tests/api/test_thermal.py [hunk 27]
```diff
@@ -206,3 +206,17 @@ def test_set_high_threshold(self, duthost, localhost, platform_api_conn):
...
+            if self.expect(position is not None, "Failed to perform get_position_in_parent for thermal {}".format(i)):
+                self.expect(isinstance(position, int), "Position value must be an integer value for thermal {}".format(i))
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_thermals):
+            replaceable = thermal.is_replaceable(platform_api_conn, i)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for thermal {}".format(i)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for thermal {}".format(i))
+        self.assert_expectations()
```
Comment (sujinmkang) [inline tests/platform_tests/api/test_thermal.py:222]: same comments as test_chassis.py.

File: tests/platform_tests/api/test_fan_drawer.py [hunk 28]
```diff
@@ -154,3 +154,17 @@ def test_set_fan_drawers_led(self, duthost, localhost, platform_api_conn):
...
+            if self.expect(position is not None, "Failed to perform get_position_in_parent for fan drawer {}".format(i)):
+                self.expect(isinstance(position, int), "Position value must be an integer value for fan drawer {}".format(i))
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        for i in range(self.num_fan_drawers):
+            replaceable = fan_drawer.is_replaceable(platform_api_conn, i)
+            if self.expect(replaceable is not None, "Failed to perform is_replaceable for fan drawer {}".format(i)):
+                self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value for fan drawer {}".format(i))
+        self.assert_expectations()
```
Comment (sujinmkang) [inline tests/platform_tests/api/test_fan_drawer.py:170]: same comments as test_chassis.py.

File: tests/platform_tests/api/test_chassis.py [hunk 29]
```diff
@@ -381,3 +381,15 @@ def test_get_watchdog(self, duthost, localhost, platform_api_conn):
...
+        position = chassis.get_position_in_parent(platform_api_conn)
+        if self.expect(position is not None, "Failed to perform get_position_in_parent"):
+            self.expect(isinstance(position, int), "Position value must be an integer value")
+        self.assert_expectations()
+
+    def test_is_replaceable(self, platform_api_conn):
+        replaceable = chassis.is_replaceable(platform_api_conn)
+        if self.expect(replaceable is not None, "Failed to perform is_replaceable"):
+            self.expect(isinstance(replaceable, bool), "Replaceable value must be a bool value")
+        self.assert_expectations()
```
Comment (sujinmkang) [inline tests/platform_tests/api/test_chassis.py:395]: get_position_in_parent() and is_replaceable() are the inherited DeviceBase class apis. Can you please move these up to DeviceBase apis test section?

**Review by jleveque — APPROVED** [review]

File: tests/common/helpers/platform_api/fan.py [hunk 30]
```diff
@@ -71,3 +71,11 @@ def set_status_led(conn, index, color):
...
 def get_status_led(conn, index):
     return fan_api(conn, index, 'get_status_led')
+
+
+def get_position_in_parent(conn, index):
+    return fan_api(conn, index, 'get_position_in_parent')
+
+
+def is_replaceable(conn, index):
+    return fan_api(conn, index, 'is_replaceable')
```
Comment (sujinmkang) [inline tests/common/helpers/platform_api/fan.py:81]: Can we also move this up to the DeviceBase class section?

File: tests/common/helpers/platform_api/fan_drawer.py [hunk 31]
```diff
@@ -59,3 +59,11 @@ def set_status_led(conn, index, color):
...
 def get_status_led(conn, index):
     return fan_drawer_api(conn, index, 'get_status_led')
+
+
+def get_position_in_parent(conn, index):
+    return fan_drawer_api(conn, index, 'get_position_in_parent')
+
+
+def is_replaceable(conn, index):
+    return fan_drawer_api(conn, index, 'is_replaceable')
```
Comment (sujinmkang) [inline tests/common/helpers/platform_api/fan_drawer.py:69]: Can we also move this up to the DeviceBase class section?

File: tests/common/helpers/platform_api/psu.py [hunk 32]
```diff
@@ -99,3 +99,22 @@ def get_voltage_high_threshold(conn, psu_id):
...
     return psu_api(conn, psu_id, 'get_voltage_low_threshold')
 
+
+def get_position_in_parent(conn, psu_id):
+    return psu_api(conn, psu_id, 'get_position_in_parent')
+
+
+def is_replaceable(conn, psu_id):
+    return psu_api(conn, psu_id, 'is_replaceable')
+
```
Comment (sujinmkang) [inline tests/common/helpers/platform_api/psu.py:109]: Can we also move this up to the DeviceBase class section?

File: tests/common/helpers/platform_api/thermal.py [hunk 33]
```diff
@@ -71,3 +71,11 @@ def get_high_critical_threshold(conn, index):
...
 def get_low_critical_threshold(conn, index):
     return thermal_api(conn, index, 'get_low_critical_threshold')
+
+
+def get_position_in_parent(conn, index):
+    return thermal_api(conn, index, 'get_position_in_parent')
+
+
+def is_replaceable(conn, index):
+    return thermal_api(conn, index, 'is_replaceable')
```
Comment (sujinmkang) [inline tests/common/helpers/platform_api/thermal.py:81]: Can we also move this up to the DeviceBase class section?

File: tests/common/helpers/platform_api/sfp.py [hunk 34]
```diff
@@ -128,3 +128,23 @@ def set_lpmode(conn, index, lpmode):
...
     return sfp_api(conn, index, 'set_power_override', [power_override, power_set])
+
+
+def get_position_in_parent(conn, index):
+    return sfp_api(conn, index, 'get_position_in_parent')
+
+
+def is_replaceable(conn, index):
+    return sfp_api(conn, index, 'is_replaceable')
+
```
Comment (sujinmkang) [inline tests/common/helpers/platform_api/sfp.py:139]: Can we also move this up to the DeviceBase class section?

File: tests/common/helpers/platform_api/chassis.py [hunk 35]
```diff
@@ -162,3 +162,11 @@ def get_watchdog(conn):
...
 def get_eeprom(conn):
     return chassis_api(conn, 'get_eeprom')
+
+
+def get_position_in_parent(conn):
+    return chassis_api(conn, 'get_position_in_parent')
+
+
+def is_replaceable(conn):
+    return chassis_api(conn, 'is_replaceable')
```
Comment (sujinmkang) [inline tests/common/helpers/platform_api/chassis.py:172]: Can we also move this up to the DeviceBase class section?

**Review by jleveque — APPROVED** [review]

**Review by vdahiya12 — APPROVED** [review]

**Review by sujinmkang — APPROVED** [review]

**Comment by jleveque** [comment]
@Junchao-Mellanox: Since these methods are inherited from DeviceBase, there are other classes which inherit these methods. Even though they may not be used by the MIB, we should test that the APIs are implemented for completeness. Do you agree? If so, can you please add similar API tests to the following files and their respective helpers?

- test_component.py
- test_fan_drawer_fans.py
- test_module.py
- test_psu_fans.py

**Comment by Junchao-Mellanox** [comment]
> @Junchao-Mellanox: Since these methods are inherited from DeviceBase, there are other classes which inherit these methods. Even though they may not be used by the MIB, we should test that the APIs are implemented for completeness. Do you agree? If so, can you please add similar API tests to the following files and their respective helpers?
> 
> * test_component.py
> * test_fan_drawer_fans.py
> * test_module.py
> * test_psu_fans.py

Agree, will do it.

**Pull request merged**